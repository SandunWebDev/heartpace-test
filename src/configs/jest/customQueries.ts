import {
	queryHelpers,
	buildQueries,
	Matcher,
	MatcherOptions,
} from '@testing-library/react';

// Custom query to directly match Cypress "data-cy" elements.
const queryAllByDataCy = (
	container: HTMLElement,
	id: Matcher,
	options?: MatcherOptions,
) => queryHelpers.queryAllByAttribute('data-testid', container, id, options);

const getMultipleError = (_c: Element | null, dataCyValue: string) =>
	`Found multiple elements with the data-test attribute of: ${dataCyValue}`;
const getMissingError = (_c: Element | null, dataCyValue: string) =>
	`Unable to find an element with the data-cy attribute of: ${dataCyValue}`;

const [
	queryByDataCy,
	getAllByDataCy,
	getByDataCy,
	findAllByDataCy,
	findByDataCy,
] = buildQueries(queryAllByDataCy, getMultipleError, getMissingError);

export {
	queryByDataCy,
	queryAllByDataCy,
	getByDataCy,
	getAllByDataCy,
	findAllByDataCy,
	findByDataCy,
};
