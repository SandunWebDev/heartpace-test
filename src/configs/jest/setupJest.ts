import '@testing-library/jest-dom';
import ResizeObserver from 'resize-observer-polyfill';

// Mocking ResizeObserver for "Recharts"
global.ResizeObserver = ResizeObserver;
