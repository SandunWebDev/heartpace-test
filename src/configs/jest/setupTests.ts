import '@testing-library/jest-dom';

// Polyfilling ResizeObserver. Currently mainly needed for "Recharts"
import ResizeObserver from 'resize-observer-polyfill';
global.ResizeObserver = ResizeObserver;
