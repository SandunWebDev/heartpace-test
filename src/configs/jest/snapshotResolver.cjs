// Making "snapshot"  files are create besides the test file.

const resolveSnapshotPath = (
	testPath, // Path of the test file being test3ed
	snapshotExtension, // The extension for snapshots (.snap usually)
) => {
	const snapshotFilePath = testPath + snapshotExtension; // Ex. some.test.tsx + '.snap'
	return snapshotFilePath;
};

const resolveTestPath = (
	snapshotFilePath, // The filename of the snapshot (i.e. some.test.tsx.snap)
	snapshotExtension, // The extension for snapshots (.snap)
) => {
	const testPath = snapshotFilePath
		.replace(snapshotExtension, '')
		.replace('__snapshots__/', ''); //Remove the .snap
	return testPath;
};

/* Used to validate resolveTestPath(resolveSnapshotPath( {this} )) */
const testPathForConsistencyCheck = 'some.test.tsx';

module.exports = {
	resolveSnapshotPath,
	resolveTestPath,
	testPathForConsistencyCheck,
};
