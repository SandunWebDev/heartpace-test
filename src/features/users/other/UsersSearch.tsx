import TextField from '@mui/material/TextField';
import PersonSearchIcon from '@mui/icons-material/PersonSearch';
import CloseSharpIcon from '@mui/icons-material/CloseSharp';
import InputAdornment from '@mui/material/InputAdornment';

export interface UsersSearchProps {
	placeholder?: string;
	value: string;
	onChange: (value: string) => void;
	onClose?: () => void;
}

export default function UsersSearch({
	placeholder = 'Search Users',
	value,
	onChange,
	onClose,
}: UsersSearchProps) {
	return (
		<TextField
			fullWidth
			size='medium'
			variant='filled'
			value={value}
			onChange={(e) => {
				onChange(e.target.value);
			}}
			placeholder={placeholder}
			slotProps={{
				input: {
					startAdornment: (
						<InputAdornment position='start'>
							<PersonSearchIcon sx={{ fontSize: '30px' }} />
						</InputAdornment>
					),
					endAdornment: (
						<InputAdornment
							data-testid='UsersSearch__closeIcon'
							position='end'
							sx={{ display: value ? 'flex' : 'none', cursor: 'pointer' }}
							onClick={onClose}>
							<CloseSharpIcon />
						</InputAdornment>
					),
				},
			}}
		/>
	);
}
