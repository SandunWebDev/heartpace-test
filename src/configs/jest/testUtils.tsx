import React, { ReactElement } from 'react';
import { render, queries, within, RenderOptions } from '@testing-library/react';
import { Provider as ReduxProvider } from 'react-redux';
import { ToastContainer } from 'react-toastify';

import MuiThemeProvider from '../theme/MuiThemeProvider';
import { store } from '../../services/redux/store';
import * as customQueries from './customQueries';

// All the default providers should goes here. So we don't have to wrap these for each test.
const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
	return (
		<MuiThemeProvider>
			<ReduxProvider store={store}>
				<ToastContainer />
				{children}
			</ReduxProvider>
		</MuiThemeProvider>
	);
};

const allQueries = {
	...queries,
	...customQueries,
};

// Re-exporting "React Testing Library's" original "render, screen, within" with our Custom Wrappers and Custom Queries.
const customRender = (
	ui: ReactElement,
	options?: Omit<RenderOptions, 'wrapper' | 'queries'>,
) => render(ui, { wrapper: AllTheProviders, queries: allQueries, ...options });
const customScreen = within(document.body, allQueries);
const customWithin = (element: HTMLElement) => within(element, allQueries);

export * from '@testing-library/react';
export {
	customRender as render,
	customScreen as screen,
	customWithin as within,
};
