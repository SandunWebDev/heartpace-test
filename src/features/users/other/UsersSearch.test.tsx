import { useState } from 'react';
import userEvent from '@testing-library/user-event';

import { render, screen, cleanup } from '../../../configs/jest/testUtils';
import UsersSearch, { UsersSearchProps } from './UsersSearch';

interface UserSearchWrapperProps extends Partial<UsersSearchProps> {
	mockOnChange: jest.Mock;
	mockOnClose: jest.Mock;
}
const UserSearchWrapper = ({
	mockOnChange,
	mockOnClose,
	...props
}: UserSearchWrapperProps) => {
	const [value, setValue] = useState('');
	mockOnChange.mockImplementation((v: string) => {
		setValue(v);
	});

	mockOnClose.mockImplementation(() => {
		setValue('');
	});

	return (
		<UsersSearch
			{...props}
			value={value}
			onChange={mockOnChange}
			onClose={mockOnClose}
		/>
	);
};

describe('UsersStatus Component', () => {
	afterEach(cleanup);

	it('snapshot', () => {
		const mockOnChange = jest.fn();
		const mockOnClose = jest.fn();

		const { asFragment } = render(
			<UserSearchWrapper
				placeholder='Search Users'
				mockOnChange={mockOnChange}
				mockOnClose={mockOnClose}
			/>,
		);
		expect(asFragment()).toMatchSnapshot();
	});

	it('should be able to type values', async () => {
		const fireUserEvent = userEvent.setup();
		const mockOnChange = jest.fn();
		const mockOnClose = jest.fn();

		render(
			<UserSearchWrapper
				placeholder='Search Users'
				mockOnChange={mockOnChange}
				mockOnClose={mockOnClose}
			/>,
		);

		const input: HTMLInputElement = screen.getByPlaceholderText('Search Users');

		await fireUserEvent.type(input, 'John');

		expect(mockOnChange).toHaveBeenCalledTimes(4);
		expect(input.value).toEqual('John');
	});

	it('should be able to clear existing value', async () => {
		const fireUserEvent = userEvent.setup();
		const mockOnChange = jest.fn();
		const mockOnClose = jest.fn();

		render(
			<UserSearchWrapper
				placeholder='Search Users'
				mockOnChange={mockOnChange}
				mockOnClose={mockOnClose}
			/>,
		);

		const input: HTMLInputElement = screen.getByPlaceholderText('Search Users');
		const closeButton: HTMLInputElement = screen.getByTestId(
			'UsersSearch__closeIcon',
		);

		// Before any value available on input.
		expect(closeButton).not.toBeVisible();

		// After some value available on input.
		await fireUserEvent.type(input, 'John');
		expect(input.value).toEqual('John');
		expect(closeButton).toBeVisible();
		expect(closeButton).toBeInTheDocument();

		// After close button is clicked
		await fireUserEvent.click(closeButton);
		expect(input.value).toEqual('');
		expect(closeButton).not.toBeVisible();
		expect(mockOnClose).toHaveBeenCalledTimes(1);
	});
});
