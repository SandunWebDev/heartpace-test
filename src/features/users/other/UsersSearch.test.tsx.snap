// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersStatus Component snapshot 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
  >
    <div
      class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiFilledInput-adornedStart MuiFilledInput-adornedEnd MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd css-174bgbr-MuiInputBase-root-MuiFilledInput-root"
    >
      <div
        class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-filled MuiInputAdornment-sizeMedium css-1sbc8xx-MuiInputAdornment-root"
      >
        <span
          aria-hidden="true"
          class="notranslate"
        >
          â€‹
        </span>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1sdz7zf-MuiSvgIcon-root"
          data-testid="PersonSearchIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <circle
            cx="10"
            cy="8"
            r="4"
          />
          <path
            d="M10.35 14.01C7.62 13.91 2 15.27 2 18v2h9.54c-2.47-2.76-1.23-5.89-1.19-5.99m9.08 4.01c.36-.59.57-1.28.57-2.02 0-2.21-1.79-4-4-4s-4 1.79-4 4 1.79 4 4 4c.74 0 1.43-.22 2.02-.57L20.59 22 22 20.59zM16 18c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2"
          />
        </svg>
      </div>
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd css-9kca0t-MuiInputBase-input-MuiFilledInput-input"
        id=":r0:"
        placeholder="Search Users"
        type="text"
        value=""
      />
      <div
        class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-filled MuiInputAdornment-sizeMedium css-138pd90-MuiInputAdornment-root"
        data-testid="UsersSearch__closeIcon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="CloseSharpIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
